modules/optimization/src/executioners/OptimizeSolve.C

```diff
@@ -0,0 +1,409 @@
+  return params;
+}
+
+OptimizeSolve::OptimizeSolve(Executioner & ex)
+  : SolveObject(ex),
+    _my_comm(MPI_COMM_SELF),
+    _solve_on(getParam<ExecFlagEnum>("solve_on")),
+    _verbose(getParam<bool>("verbose")),
+    _tao_solver_enum(getParam<MooseEnum>("tao_solver").getEnum<TaoSolverEnum>()),
+    _parameters(std::make_unique<libMesh::PetscVector<Number>>(_my_comm))
+{
+  if (libMesh::n_threads() > 1)
+    mooseError("OptimizeSolve does not currently support threaded execution");
+}
+
+bool
+OptimizeSolve::solve()
+{
+  // Initial solve
+  _inner_solve->solve();
+
+  // Grab objectivefunction
+  if (!_problem.hasUserObject("OptimizationReporter"))
+    mooseError("No OptimizationReporter object found.");
+  _obj_function = &_problem.getUserObject<OptimizationReporter>("OptimizationReporter");
+
+  // Initialize solution and matrix
+  _obj_function->setInitialCondition(*_parameters.get());
+  _ndof = _parameters->size();
+  bool solveInfo = (taoSolve() == 0);
+  return solveInfo;
+}
+
+PetscErrorCode
+OptimizeSolve::taoSolve()
+{
+  // Petsc error code to be checked after each petsc call
+  PetscErrorCode ierr = 0;
+
+  // Initialize tao object
+  ierr = TaoCreate(_my_comm.get(), &_tao);
+  CHKERRQ(ierr);
+
+  TaoSetMonitor(_tao, monitor, this, nullptr);
+
+  switch (_tao_solver_enum)
+  {
+    case TaoSolverEnum::NEWTON_TRUST_REGION:
+      ierr = TaoSetType(_tao, TAONTR);
+      break;
+    case TaoSolverEnum::BOUNDED_NEWTON_TRUST_REGION:
+      ierr = TaoSetType(_tao, TAOBNTR);
+      break;
+    case TaoSolverEnum::BOUNDED_CONJUGATE_GRADIENT:
+      ierr = TaoSetType(_tao, TAOBNCG);
+      break;
+    case TaoSolverEnum::NEWTON_LINE_SEARCH:
+      ierr = TaoSetType(_tao, TAONLS);
+      break;
+    case TaoSolverEnum::BOUNDED_NEWTON_LINE_SEARCH:
+      ierr = TaoSetType(_tao, TAOBNLS);
+      break;
+    case TaoSolverEnum::NEWTON_TRUST_LINE:
+      ierr = TaoSetType(_tao, TAONTL);
+      break;
+    case TaoSolverEnum::BOUNDED_NEWTON_TRUST_LINE:
+      ierr = TaoSetType(_tao, TAOBNTL);
+      break;
+    case TaoSolverEnum::QUASI_NEWTON:
+      ierr = TaoSetType(_tao, TAOLMVM);
+      break;
+    case TaoSolverEnum::BOUNDED_QUASI_NEWTON:
+      ierr = TaoSetType(_tao, TAOBLMVM);
+      break;
+
+    case TaoSolverEnum::NELDER_MEAD:
+      ierr = TaoSetType(_tao, TAONM);
+      break;
+
+    case TaoSolverEnum::BOUNDED_QUASI_NEWTON_LINE_SEARCH:
+      ierr = TaoSetType(_tao, TAOBQNLS);
+      break;
+    case TaoSolverEnum::ORTHANT_QUASI_NEWTON:
+      ierr = TaoSetType(_tao, TAOOWLQN);
+      break;
+    case TaoSolverEnum::GRADIENT_PROJECTION_CONJUGATE_GRADIENT:
+      ierr = TaoSetType(_tao, TAOGPCG);
+      break;
+    case TaoSolverEnum::BUNDLE_RISK_MIN:
+      ierr = TaoSetType(_tao, TAOBMRM);
+      break;
+
+    default:
+      mooseError("Invalid Tao solve type");
+  }
+
+  CHKERRQ(ierr);
+  // Set bounds for bounded optimization
+  ierr = TaoSetVariableBoundsRoutine(_tao, variableBoundsWrapper, this);
```

Comment:
you could return an error here for all the solvers than dont accept bounds?

Reply:
That would be a different behavior from the TaoSetVariableBounds interface.  It mostly checks that the size is correct.